### PR TITLE
[ALOY-86] Resolve widgets as NodeJS modules

### PR DESCRIPTION
**Pull request by FokkeZB**
- Resolve NPM packages as Alloy widgets (ALOY-86)
- Require NPM packages to have `alloy-widget-` prefix
- Require NPM packages to have certain keywords
- Report recursive dependencies that are missing

https://jira.appcelerator.org/browse/ALOY-86

@feons how do add a test app that includes files (`package.json`, `node_modules`) on the project-level? All test apps start at the `app` folder level.

### CODE REVIEW

**Comment by FokkeZB** [comment]
The logs should look like (default, local and global used here):

```
[INFO] [widget.xml] foo view processing...
[INFO]   style:      "widget.tss"
[INFO]   view:       "widget.xml"
[INFO]   controller: "widget.js"
[TRACE] WIDGET_SRC=node_modules/foo/assets
[TRACE]
[TRACE] - Processing "builtins" module...
[TRACE] - Processing "optimizer" module...
[TRACE] - Processing "compress" module...
[INFO]   created:    "Resources/iphone/alloy/widgets/foo/controllers/widget.js"
[DEBUG]   map:        "build/map/Resources/iphone/alloy/widgets/foo/controllers/widget.js.map"
[INFO]   created:     "Resources/iphone/alloy/widgets/foo/styles/widget.js"
[INFO] [widget.xml] bar view processing...
[INFO]   style:      "widget.tss"
[INFO]   view:       "widget.xml"
[INFO]   controller: "widget.js"
[TRACE] WIDGET_SRC=app/widgets/bar/assets
[TRACE]
[TRACE] - Processing "builtins" module...
[TRACE] - Processing "optimizer" module...
[TRACE] - Processing "compress" module...
[INFO]   created:    "Resources/iphone/alloy/widgets/bar/controllers/widget.js"
[DEBUG]   map:        "build/map/Resources/iphone/alloy/widgets/bar/controllers/widget.js.map"
[INFO]   created:     "Resources/iphone/alloy/widgets/bar/styles/widget.js"
[INFO] [widget.xml] zar view processing...
[INFO]   style:      "widget.tss"
[INFO]   view:       "widget.xml"
[INFO]   controller: "widget.js"
[TRACE] WIDGET_SRC=../../../../../usr/local/lib/node_modules/zar/assets
```

File: Alloy/utils.js [hunk 1]
```diff
@@ -306,52 +308,90 @@ exports.getWidgetDirectories = function(appDir) {
...
+			collection = findWidgetAsNodeModule(id);
 
-        dirs.push(collection);
+			if (!collection) {
+				notFound.push(id);
+				return;
+			}
+		}
+
+    dirs.push(collection);
```
Comment (feons) [inline Alloy/utils.js]: Need to fix the indentation.

File: Alloy/utils.js [hunk 2]
```diff
@@ -306,52 +308,90 @@ exports.getWidgetDirectories = function(appDir) {
...
 	// walk the dependencies, tracking any missing widgets
 	var notFound = [];
-    for (var id in appWidgets) {
-		if (!collections[id]) {
-			notFound.push(id);
-		} else {
-			walkWidgetDependencies(collections[id]);
-		}
-    }
+  for (var id in appWidgets) {
```
Comment (feons) [inline Alloy/utils.js]: Fix indentation.

File: Alloy/utils.js [hunk 3]
```diff
@@ -306,52 +308,90 @@ exports.getWidgetDirectories = function(appDir) {
...
+		walkWidgetDependencies(id);
+	}
 
     // if there are missing widgets, abort and tell the developer which ones
-    if (!!notFound.length) {
+  if (!!notFound.length) {
```
Comment (feons) [inline Alloy/utils.js]: Fix indentation.

**Comment by feons** [comment]
@FokkeZB, you would have to modify https://github.com/appcelerator/alloy/tree/master/test/projects/HarnessTemplate to include files at the project level.

**Comment by FokkeZB** [comment]
@feons test added. Can't think of a way to test globally installed NPM module but tested that locally and the only difference is that it will come from yet another path, which would have no different effect then local NPM modules.

**Comment by FokkeZB** [comment]
Anything you need here to move on @feons ?

**Comment by feons** [comment]
@FokkeZB, I'm very sorry for the delayed response. The failing test definitely need some love.

**Comment by FokkeZB** [comment]
@feons, can I get details on why it failed? The test runs fine for me if I run it local:

```
$ app=testing/ALOY-86 jake test:spec[compile.js]
Loading test spec from "test/specs/compile.js"
Begin Alloy Test Suite
[PASS] alloy compile testing/ALOY-86 --> preparing test app
[PASS] alloy compile testing/ALOY-86 [android] --> compiles without critical error
[PASS] alloy compile testing/ALOY-86 [android] --> leaves no compiler directives in generated code
[PASS] alloy compile testing/ALOY-86 [android] --> has no undefined style entries
[PASS] alloy compile testing/ALOY-86 [android] --> generated a alloy/controllers/index.js file
[PASS] alloy compile testing/ALOY-86 [android] --> matches known good generated code for alloy/controllers/index.js
[PASS] alloy compile testing/ALOY-86 [android] --> generated a alloy/widgets/classic/controllers/widget.js file
[PASS] alloy compile testing/ALOY-86 [android] --> matches known good generated code for alloy/widgets/classic/controllers/widget.js
[PASS] alloy compile testing/ALOY-86 [android] --> generated a alloy/widgets/classic/styles/widget.js file
[PASS] alloy compile testing/ALOY-86 [android] --> matches known good generated code for alloy/widgets/classic/styles/widget.js
[PASS] alloy compile testing/ALOY-86 [android] --> generated a alloy/widgets/npm/controllers/widget.js file
[PASS] alloy compile testing/ALOY-86 [android] --> matches known good generated code for alloy/widgets/npm/controllers/widget.js
[PASS] alloy compile testing/ALOY-86 [android] --> generated a alloy/widgets/npm/styles/widget.js file
[PASS] alloy compile testing/ALOY-86 [android] --> matches known good generated code for alloy/widgets/npm/styles/widget.js
[PASS] alloy compile testing/ALOY-86 [ios] --> compiles without critical error
[PASS] alloy compile testing/ALOY-86 [ios] --> leaves no compiler directives in generated code
[PASS] alloy compile testing/ALOY-86 [ios] --> has no undefined style entries
[PASS] alloy compile testing/ALOY-86 [ios] --> generated a alloy/controllers/index.js file
[PASS] alloy compile testing/ALOY-86 [ios] --> matches known good generated code for alloy/controllers/index.js
[PASS] alloy compile testing/ALOY-86 [ios] --> generated a alloy/widgets/classic/controllers/widget.js file
[PASS] alloy compile testing/ALOY-86 [ios] --> matches known good generated code for alloy/widgets/classic/controllers/widget.js
[PASS] alloy compile testing/ALOY-86 [ios] --> generated a alloy/widgets/classic/styles/widget.js file
[PASS] alloy compile testing/ALOY-86 [ios] --> matches known good generated code for alloy/widgets/classic/styles/widget.js
[PASS] alloy compile testing/ALOY-86 [ios] --> generated a alloy/widgets/npm/controllers/widget.js file
[PASS] alloy compile testing/ALOY-86 [ios] --> matches known good generated code for alloy/widgets/npm/controllers/widget.js
[PASS] alloy compile testing/ALOY-86 [ios] --> generated a alloy/widgets/npm/styles/widget.js file
[PASS] alloy compile testing/ALOY-86 [ios] --> matches known good generated code for alloy/widgets/npm/styles/widget.js
[PASS] alloy compile testing/ALOY-86 [mobileweb] --> compiles without critical error
[PASS] alloy compile testing/ALOY-86 [mobileweb] --> leaves no compiler directives in generated code
[PASS] alloy compile testing/ALOY-86 [mobileweb] --> has no undefined style entries
[PASS] alloy compile testing/ALOY-86 [mobileweb] --> generated a alloy/controllers/index.js file
[PASS] alloy compile testing/ALOY-86 [mobileweb] --> matches known good generated code for alloy/controllers/index.js
[PASS] alloy compile testing/ALOY-86 [mobileweb] --> generated a alloy/widgets/classic/controllers/widget.js file
[PASS] alloy compile testing/ALOY-86 [mobileweb] --> matches known good generated code for alloy/widgets/classic/controllers/widget.js
[PASS] alloy compile testing/ALOY-86 [mobileweb] --> generated a alloy/widgets/classic/styles/widget.js file
[PASS] alloy compile testing/ALOY-86 [mobileweb] --> matches known good generated code for alloy/widgets/classic/styles/widget.js
[PASS] alloy compile testing/ALOY-86 [mobileweb] --> generated a alloy/widgets/npm/controllers/widget.js file
[PASS] alloy compile testing/ALOY-86 [mobileweb] --> matches known good generated code for alloy/widgets/npm/controllers/widget.js
[PASS] alloy compile testing/ALOY-86 [mobileweb] --> generated a alloy/widgets/npm/styles/widget.js file
[PASS] alloy compile testing/ALOY-86 [mobileweb] --> matches known good generated code for alloy/widgets/npm/styles/widget.js
Finished in 4.657 seconds
40 specs, 0 failures
```

**Comment by feons** [comment]
You can click on `Show all checks` below then go to the Travis CI build details page to view build log. 
Looks like the test failed because the npm module `alloy-widget-npm` is not installed. You can add it to  `HarnessTemplate/node_modules/`.

**Comment by FokkeZB** [comment]
Ah... of course, the `node_modules` folder was ignored. I've added it and it's running tests now.

**Comment by FokkeZB** [comment]
@feons all green!

**Comment by feons** [comment]
**Approved.**

**Pull request merged**